44 consolidate input components to input set

ianmarquez opened the pull request:
- still missing hook for overriding styles
- still missing dropdown component

missing functionalities are still in progress and are not yet ready to be merged in, will work in a separate branch for those

mariatorrentedev commented on packages/components/src/form/FieldSet.tsx:
```diff
@@ -0,0 +1,102 @@
+  children,
+  title,
+}: Props) => {
+  const cn: FieldSetStyles = {
+    label: cns('font-bold text-sky-500 text-sm', customStyles?.label),
+    root: cns(
+      'p-4 flex flex-col gap-2 bg-gray-800 text-xl',
+      customStyles?.root
+    ),
+    fieldset: cns(
+      'grid grid-cols-2 gap-2 disabled:cursor-disabled',
+      customStyles?.fieldset
+    ),
+    title: cns(customStyles?.title),
+  }
```
@ianmarquez there is a method called `appliedStyles` that can be exported from `@stcland/utils` which does the deep merging with `default` and `customStyles`, so that we don't need to use the `cns` per prop, can we update the occurrences that used this approach?

See example in `Table`, `Button` or `Icon`

mariatorrentedev commented on packages/components/src/form/FieldSet.tsx:
```diff
@@ -0,0 +1,102 @@
+  children: React.ReactNode[];
+}
+
+export type PanelStyles = Partial<{
+  root: React.HTMLAttributes<HTMLDivElement>['className'];
+  title: React.HTMLAttributes<HTMLDivElement>['className'];
+}>;
```
same here.

mariatorrentedev commented on packages/components/src/form/Form.tsx:
```diff
@@ -0,0 +1,130 @@
+  defaultValues?: { [key: string]: any };
+  title?: string;
+  withPanel?: boolean;
+  panelClassName?: string; // applied to panel
+  formClassName?: string; // applied to form element
```
@ianmarquez should we merged this two props as customStyles for the Form component? What do you think? and also maybe include a root, for the outer wrapper.

mariatorrentedev commented on packages/components/src/form/FormInput.tsx:
```diff
@@ -0,0 +1,200 @@
+  onChange?: () => void;
+};
+
+export const formInputDefaultProps = {
+  label: '',
+  placeholder: '',
+  valueAsNumber: false,
+  required: false,
+  readOnly: false,
+  disabled: false,
+  hidden: false,
+  touched: false,
+  dirty: false,
+  onFocus: function () {},
+  textAreaProps: {},
+}
```
@ianmarquez should we delete this variable? I don't see that we are using it anywhere, maybe an oversight? considering that we pass the defaults in the component signature.

mariatorrentedev commented on packages/components/src/form/FormInput.tsx:
```diff
@@ -0,0 +1,200 @@
+    message: 'is required',
+  }
+
+  const inputStyles: FormInputStyles = {
+    root: cns('flex flex-col w-100 gap-2', customStyles?.root),
+    label: cns('text-sm', customStyles?.label),
+    input: cns(
+      type !== FieldType.TEXTAREA && 'h-14',
+      'px-4 py-4 border-[#4B5563] border-solid rounded-lg bg-[#2A2C30] text-lg border text-white w-full',
+      'placeholder:italic',
+      'disabled:text-zinc-400',
+      customStyles?.input
+    ),
+    error: cns('text-red-300 italic text-sm', customStyles?.error),
+  }
```
same here.

mariatorrentedev commented on packages/components/src/form/FormInput.tsx:
```diff
@@ -0,0 +1,200 @@
+          {...{ labelText: label, required }}
+        />
+      )}
+      <div className="relative">
```
should this style be part of the `inputStyles` something like `inputRoot`?

mariatorrentedev replied:
I am talking about the `className="relative"` we should be able to add it to the inputStyles, and make it available for the consumer.

ianmarquez replied:
exposed this with inputContainer prop

mariatorrentedev commented on packages/components/src/form/FormLabel.tsx:
```diff
@@ -0,0 +1,38 @@
+  labelText,
+  className,
+  required,
+  requiredClass,
```
should we change this prop to be `customStyles`? and also let users update either `root` or `required`?

mariatorrentedev commented on packages/components/src/form/inputs/InputSet.tsx:
This is such an elegant work @ianmarquez! Love the composition done here `InputSet.tsx` 🎊 🙌

mariatorrentedev commented on packages/components/src/form/inputs/RadioButtonGroup.tsx:
```diff
@@ -0,0 +1,93 @@
+  defaultChecked?: string;
+  radioButtons: RadioButtonShape[];
+  readOnly?: boolean;
+  required: boolean;
```
Why is `required` the only required prop(LOL)? if we are setting a default value?

ianmarquez replied:
radioButtons (the options) is also required so is name

mariatorrentedev commented on packages/components/src/form/inputs/RadioButtonGroup.tsx:
```diff
@@ -0,0 +1,93 @@
+  }
+  const flexDir = direction === 'horizontal' ? 'flex-row' : 'flex-col'
+
+  const cn = {
+    root: cns('mb-2 flex gap-4', className),
+    buttonContainer: cns('flex', flexDir),
+    label: 'text-white/50',
+    radioContainer: 'flex gap-2',
+    radioButton: (row: number) =>
+      cns(
+        'accent-app-primary-500',
+        direction === 'horizontal' && row !== 0 && 'flex-3',
+      ),
+  }
```
I think we should create a `RadioButtonGroupStyles`, and also pass a `customStyles` prop... so we can allow dynamic styles from consumer perspective.

mariatorrentedev commented on packages/components/src/form/inputs/Select.tsx:
```diff
@@ -0,0 +1,109 @@
+  onSelectionChange?: (value: string) => void;
+  className?: string;
+  disabled?: boolean;
+  classNames?: {
+    menuButton?: string;
+    label?: string;
+  };
```
Would love to have your input, but I think we should update this `classNames` property to be `customStyles` for consistency, and include:
- `listItem`.
- `color` for `primaryColor`

ianmarquez replied:
i think changing it to customStyles would be better, just so its uniform across all components makes it easier for the consumers of the library as well

mariatorrentedev commented on packages/storybook/stories/InputDate.mdx:
```diff
@@ -0,0 +1,38 @@
+<Canvas of={DateInputStories.Hidden} />
+#### Touched
+<Canvas of={DateInputStories.Touched} />
+#### Dirty
+<Canvas of={DateInputStories.Dirty} />
+---
```
@ianmarquez  How can we test the `dirty` state? Hmm should we use the boolean in `FormInput` in combination with the `error` style?

ianmarquez replied:
so far we cannot test as of yet, the dirty state and touched state is not doing anything for now. Its just there to determine if field is touched or dirty. This state comes from the react-hook-forms framework. Im still in the midst of implementing it although its just something thats nice to have.

mariatorrentedev commented on packages/storybook/stories/InputDate.stories.tsx:
```diff
@@ -0,0 +1,119 @@
+  hidden: true,
+}
+
+export const Touched = DateInputTemplate.bind({})
+Touched.args = {
+  name: 'dateInput',
+  placeholder: 'DateInput',
+  label: 'DateInput Input',
+  touched: true,
+}
```
Same with `touched`? what is the expectative from a user perspective in the UI?

ianmarquez replied:
touched and dirty are props injected by react-hook-form, usually its used to change state depending on the field. Since we dont have a design for it yet I havent implemented any change.

ianmarquez commented:
thanks @mariatorrentedev apologies wasnt able to make progress much during the weekdays, currently addressing your comments. Will create a new commit with all the context of overriding styles as well. Hopefully this week I get to finish everything

mariatorrentedev commented on packages/components/src/form/inputs/Select.tsx:
```diff
@@ -0,0 +1,109 @@
+  options,
+  onSelectionChange,
+  className = '',
+  classNames = {},
```
@ianmarquez can we update also this classNames to `customStyles`?

mariatorrentedev commented on packages/components/src/form/inputs/Select.tsx:
```diff
@@ -0,0 +1,109 @@
+        value={value}
+        onChange={handleChange}
+        isDisabled={disabled}
+        primaryColor="text-primary-main"
```
@ianmarquez here I think we should pass a prop to the wrapper SelectMenuInput, so that the user can pass `color` prop, or add it to the defaultStyles...I will let you choose either way... `Button` is a good referece.

mariatorrentedev replied:
What do you think about this question @ianmarquez?

mariatorrentedev commented on packages/components/src/styles/componentTypes.ts:
```diff
@@ -1,48 +1,93 @@
+}
+
+export interface FieldSetStyles extends BaseStyles {
+  label: React.HTMLAttributes<HTMLDivElement>['className'];
```
can we create type as `ClassName = React.HTMLAttributes<HTMLDivElement>['className']` and use it in all ocurrencies? I think it would be cleaner to read, just a preference.

mariatorrentedev reviewed (commented):
@ianmarquez thanks for the updates, left couple of questions/comments :)!

ianmarquez commented:
addressed comments and wired all form styleable components with the withStylesWrapper

mariatorrentedev commented on packages/components/src/form/FieldSet.tsx:
```diff
@@ -0,0 +1,82 @@
+      }}
+    >
+      {label && <div className={cn.label}>{label}</div>}
+      <fieldset {...fieldSetProps} className={cn.fieldset}>
```
@ianmarquez minor thing here, if we are passing already the `className` as part of the `fieldSetProp` above, we don't need it here, right?

mariatorrentedev commented on packages/components/src/form/FormInput.tsx:
```diff
@@ -0,0 +1,184 @@
+    message: 'is required',
+  }
+
+  const inputStyles: FormInputStyles = appliedStyles(
+    {
+      root: 'flex flex-col w-100 gap-2',
+      label: 'text-sm pt-2',
+      inputRoot: cns(
+        type !== FieldType.TEXTAREA && 'h-14',
+        'px-4 py-4 border-[#4B5563] border-solid rounded-lg bg-[#2A2C30] text-lg border text-white w-full dark:[color-scheme:dark]',
+        'disabled:text-zinc-400'
+      ),
+      error: 'text-red-300 italic text-sm',
+      info: '',
+    },
+    customStyles
+  )
```
@ianmarquez couple of things here:

1. Can we use Tailwind variants instead of hex? `#4B5563 = gray-600` and for `#2A2C30` try using any similar from Tailwind color variants?

2. What would happen if `inputRoot` does not add a :disabled pseudo-class, adds a `replace` word at the beginning OR there are not default styles? The UI design would not be as expected... I think that we should have a disabled prop as part of the inputStyles, and add the pseudo class after the styles are merged:

`cn = {
  root: mergedStyles.root,
  label: mergedStyles.root,
  inputRoot: cns(mergedStyles.inputRoot, disabled:mergedStyles.disabled),
  ...rest
}`

Can you also add customStyles in the `stc.config.ts` of Storybook, so we can test all the `withStylesWrapper` variants with customizations?

ianmarquez replied:
1st  and 3rd ask is done, however for the last one wouldn't that be the intended effect? the consumer can change everything that we expose, or are you seeing this as more of a guard rails thing wherein a consumer can only change very specific values.

mariatorrentedev commented on packages/components/src/form/inputs/RadioButtonGroup.tsx:
```diff
@@ -0,0 +1,121 @@
+    value: required,
+    message: ' is required',
+  }
+  const flexDir = direction === 'horizontal' ? 'flex-row' : 'flex-col'
+
+  const cn: RadioButtonGroupStyles = appliedStyles(
+    {
+      root: 'mb-2 flex gap-2 flex-col',
+      buttonContainer: cns('flex', flexDir, 'gap-2'),
+      label: '',
+      radioContainer: 'flex gap-2',
+      error: 'text-red-300 italic text-sm',
+      radioButton: (row: number) =>
+        cns(
+          'accent-sky-500 h-4 w-4',
+          direction === 'horizontal' && row !== 0 && 'flex-3'
+        ),
+    },
+    customStyles
+  )
```
@ianmarquez same concern here as above, there is a case where the customStyles can completely override the `buttonContainer` and we might lose the `flexDir` and `direction` style, this should be added after the styles are `merged`.

`Button` is a good example.

mariatorrentedev commented on packages/components/src/form/inputs/Select.tsx:
```diff
@@ -0,0 +1,116 @@
+      root: 'flex flex-col gap-2',
+      label: '',
+      menuButton:
+        'flex px-4 py-1 border-[#4B5563] border-solid rounded-lg bg-[#2A2C30] text-lg border text-white w-full items-center [&>*:first-child]:grow',
```
@ianmarquez same here about the hex colors, try to use variants from Tailwind colors.

mariatorrentedev commented on packages/components/src/form/withStylesWrapper.tsx:
Really great idea with this component @ianmarquez love it 🥇! Indeed we could use it for all the other `withStyles` occurrences outside of `form` right? so maybe it would be great to put it in a parent directory.

mariatorrentedev commented on packages/components/src/index.ts:
```diff
@@ -1,15 +1,62 @@
 
 // Styles
-export { StcStylesProvider, useStyles, getStyles } from './styles'
-export type { StclandStyles, TableStyles } from './styles'
+export { StcStylesProvider, useStyles, getStyles } from "./styles";
+export type { StclandStyles, TableStyles } from "./styles";
 
 // Components
-export type { TableProps } from './table'
-export { Table } from './table'
+export type { TableProps } from "./table";
+export { Table } from "./table";
 
```
@ianmarquez  Minor thing here about the ESLINT config... is not working correctly with the eslint config file? is just that we add some rules requested by Steve, and one it is `"quotes": ["error", "single"]` located in the `config/eslint.base.json` so we can have a consistent formatting 😬

mariatorrentedev commented on packages/components/src/styles/componentTypes.ts:
```diff
@@ -1,63 +1,123 @@
-    brightenOnHover: string
-    bright:  string
-  }
+export interface BaseStyles {
+  root: DivElementClassName;
+}
```
@ianmarquez great idea here to group shared types!

mariatorrentedev commented on packages/components/src/styles/defaults.ts:
```diff
@@ -1,75 +1,78 @@
+  submitButton?: Partial<ButtonStyles>;
+  resetButton?: Partial<ButtonStyles>;
+  icon?: Partial<IconStyles>;
+  accordion?: Partial<AccordionStyles>;
+};
```
This is still a WIP @ianmarquez?

mariatorrentedev commented on packages/storybook/stc.config.ts:
```diff
@@ -62,6 +64,12 @@ const styles : StclandStyles = {
+  },
+  form: {},
+>>>>>>> 3f73eae (Feature: Implement form with primitive input types)
 }
```
Wopsi @ianmarquez minor merged conflict here lol.

mariatorrentedev commented on packages/storybook/stories/Form.stories.tsx:
```diff
@@ -0,0 +1,265 @@
+  debug: true,
+}
+
+export const WithoutPanel = FieldsetTemplate.bind({})
+WithoutPanel.args = {
+  title: 'Form Title',
+  withPanel: false,
+}
```
@ianmarquez we might be missing here a default style for the background.

ianmarquez replied:
this is intended as the panel is whats giving the background color. if none it would inherit its parents background.

mariatorrentedev commented on packages/storybook/stories/Form.stories.tsx:
```diff
@@ -0,0 +1,265 @@
+  withPanel: true,
+}
+
+export const Debug = FieldsetTemplate.bind({})
+Debug.args = {
+  title: 'Form With Debug Enabled',
+  label: 'Default Fieldset Label',
+  withPanel: true,
+  debug: true,
+}
```
@ianmarquez Oh wow, really cool the debugger! 🎊

mariatorrentedev commented on packages/storybook/stories/Form.stories.tsx:
```diff
@@ -0,0 +1,265 @@
+        { value: 'option 4', label: 'Option 4' },
+      ]}
+    />
+    <CheckBox name="agree" label="Agree to Terms and Conditions" required />
```
<img width="367" alt="Screenshot 2024-11-01 at 9 27 46 AM" src="https://github.com/user-attachments/assets/78d5c8b2-3d9e-4d9e-990d-9f94cb13341f">

@ianmarquez minor thing here, but can we troubleshoot the style so that the error is rendered below the input?

mariatorrentedev reviewed (commented):
@ianmarquez love the idea of the `withStylesWrapper`! Left couple of questions/comments...let me know if you have any doubt... Really great work, oh and also the debbuger, pretty cool 😎 🙌
